src/Features/Core/Portable/CodeRefactorings/CodeRefactoringService.cs

```diff
@@ -89,18 +89,58 @@ static ProjectCodeRefactoringProvider.ExtensionInfo GetExtensionInfo(ExportCodeR
         CodeActionOptionsProvider options,
         CancellationToken cancellationToken)
     {
-        foreach (var provider in GetProviders(document))
-        {
-            cancellationToken.ThrowIfCancellationRequested();
+        // A token for controlling the inner work we do calling out to each provider.  Once we have a single provider
+        // that returns a refactoring, we can cancel the work we're doing with all other providers.
+        using var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+
+        // This await will not complete until all providers have been called (though we'll attempt to bail out of any of
+        // them once we have a single refactoring found).  So there's no concern here about produceItems running after
+        // linkedTokenSource has been diposed.
+        return await ProducerConsumer<VoidResult>.RunParallelAsync(
+            source: this.GetProviders(document),
+            produceItems: static async (provider, callback, args, cancellationToken) =>
+            {
+                var (@this, document, state, options, linkedTokenSource) = args;
 
-            var refactoring = await GetRefactoringFromProviderAsync(
-                document, state, provider, options, cancellationToken).ConfigureAwait(false);
+                // Do no work if either the outer request canceled, or another provider already found a refactoring.
+                if (cancellationToken.IsCancellationRequested || linkedTokenSource.Token.IsCancellationRequested)
+                    return;
 
-            if (refactoring != null)
-                return true;
-        }
+                try
+                {
+                    // We want to pass linkedTokenSource.Token here so that we can cancel the inner operation once the
+                    // outer ProducerConsumer sees a single refactoring returned by any provider.
+                    var refactoring = await @this.GetRefactoringFromProviderAsync(
+                        document, state, provider, options, linkedTokenSource.Token).ConfigureAwait(false);
```

Comment:
>linkedTokenSource.Token [](http://example.com/codeflow?start=60&length=23)

Is it possible to get a ObjectDisposedException here if this has been disposed?

Reply:
It should not be possible we only dispose *after* the call to RunParallelAsync has completed. And it only completes Asher both the producer and consumer complete.

Reply:
took a minute to process, but yeah, that makes sense. Thanks!